src/python/pants/backend/python/rules/coverage.md

```diff
@@ -0,0 +1,27 @@
+## How Python Coverage Works
+- running pytest generates data and drops a sql file, `.coverage` in the chroot.
+- running `coverage combine` on a list of `.coverage` files merges them into a single `.coverage` sql file.
+- running `coverage html` or `coverage xml` on a single `.coverage` file generates a report.
+
+
+## Pants Architecture for Coverage
+- Leave specific coverage options (e.g XML vs. Json output) to each language through a subsystem for that language’s coverage provider, e.g. pytest-cov
+- Core test gets a `--coverage` option which just turns coverage on and off. It also writes down the coverage reports via `workspace.materialize_directory` It should probably also log something like
+  `Python coverage report saved at coverage/python.xml`
+  `Junit coverage report saved at coverage/junit.xml`
+- Add a new concrete dataclass called something like `CoverageResult` or `MergedCoverageResult`, along with using a union so that languages provide their own implementation to create the standardized `CoverageResult` when test `--coverage` is true
+- Languages create their own rules for running coverage. For python this will likely be:
+  - `pytest-coverage-config` - new rule to generate a coveragerc file.
+  - `python-test-runner` - update exising rule to turn on pytest-cov and inject the coveragerc file into the chroot.
+  - `coverage-merge` - new rule merges the coverage data of individual test runs, also uses the coveragerc file. Outputs a merged data file.
+  - `coverage-report` - new rule generates the xml/html report as a `CoverageResult`
+
+
+## End result
+`./pants test --coverage some/package:: some/other/package::` results in one coverage report per language dropped in some reasonable location.
```

Comment:
Sorry, I'm still a little hung up on the per-language piece here... It seems like aiming for a single unified output would be nicer than splitting this up per-language if possible. In particular, if we're modifying `TestResult` to pass these things around, it feels like that should be a cross-language concept.

I don't have enough background here to strongly advocate for Cobertura XML as the one true format, but it does seem more widespread and common than any others. A quick googling showed that there are tools for manipulating and merging Cobertura XML files (e.g. https://github.com/danielpalme/ReportGenerator or https://www.npmjs.com/package/cobertura-merge), so... What are the reasons we actively _don't_ want to aim to have a single format for all languages?

Reply:
> What are the reasons we actively don't want to aim to have a single format for all languages?

There are no reasons for this. I would be happy if this is possible. My concern here is that we would like to have some form of python coverage working and in this weeks release. If it is possible to meaningfully merge coverage reports across languages I see no reason we cannot change that behavior in the future when we support coverage in V2 for any other language.

Reply:
I guess as long as we're viewing the v2 `TestResult` interface as completely unstable, it's a fine way to experiment, but I don't love adding language-specific concepts to very general interfaces... At the very least, let's call this field on `TestResult` `python_sqlite_coverage_file` or something similar to make it abundantly clear what it is (and to force anyone who consumes it to change their code when we change format).

Reply:
> At the very least, let's call this field on TestResult python_sqlite_coverage_file or something similar to make it abundantly clear what it is (and to force anyone who consumes it to change their code when we change format).

But, right now, the field `TestResult.coverage_digest` is completely generalizable. It's up to the test runner implementation to decide what files they want to store that in. `pytest-cov` might store it as a `sqlite` file because that's how that tool works, whereas a JVM coverage tool might store it in XML because that's what's idiomatic to the tool.

The only thing that `rules/core/test.py` will do with the field is to collect all the generic `TestResult.coverage_digest` and then, through a union, convert that collection into one single `CoverageReport`. _The logic for merging is provided by the language_ based on what is idiomatic for that tool, not through some generic merge rule in `rules/core`.

> I guess as long as we're viewing the v2 TestResult interface as completely unstable

I think this is true. For example, we fairly dramatically changed three weeks ago how `LintResult` and `FmtResult` work.

Reply:
> But, right now, the field TestResult.coverage_digest is completely generalizable.

But not in any useful way, right? No rule can consume it and decide what to do with it, without having a bunch more context? If the field were a `Union` with members like `python_sqlite` and `cobertura_xml`, rather than a `Digest`, I'd be ok with that as an interim solution too. Or maybe even if we had a union of `PythonTestResult` and `JavaTestResult`, though I'm still not sold on that. But right now, a rule which consumes a `TestResult` can't possibly know what to do with the field. Other than assuming that's it's a sqlite file produced from python, which is why I'd be ok with naming the field `python_sqlite_coverage_file` - it states what the Digest _is_.

Reply:
That seems fair. I had not considered which other rules might eventually handle TestResult. 

I think for now this should be something like `_ python_sqlite_coverage_file` or `experimental_ python_sqlite_coverage_file`

Reply:
> The only thing that rules/core/test.py will do with the field is to collect all the generic TestResult.coverage_digest and then, through a union, convert that collection into one single CoverageReport. The logic for merging is provided by the language based on what is idiomatic for that tool, not through some generic merge rule in rules/core.

> But not in any useful way, right?

I think that this is still useful. We'd be exposing _some_ semantic information about the content of the data (we know that it is coverage-related data). And you _can_ actually do relatively useful things with this by inspecting file contents.

Reply:
A third option somewhere between "mostly untyped digest" and "language specific special field", would be to have the field contain an optional typed wrapper, like an enum. Approximately:
```
@union
class CoverageOutput(ABC):
  @abstractproperty
  def digest(self) -> Digest: ...

@dataclass(frozen=True)
class CoberturaCoverageOutput(CoverageOutput):
  digest: Digest

@dataclass(frozen=True)
class TestResult:
  status: Status
  stdout: str
  stderr: str
  coverage: Optional[CoverageOutput] = None
```

The caller would then try to merge all coverage outputs of a particular type using the union.
This might also be premature optimization though!